docs/redconf.8.scd
redconf(8)  "Redpesk OS"

# NAME

redconf - RedNode configuration manager

# SYNOPSIS

*redconf* [<opts>] command [<args>]

# DESCRIPTION

*redconf* is an administrative tool for creating or inspecting rednodes.

The commands are:

- *create* for creating a rednode,
- *config* for getting specific rednode config,
- *mergeconfig* for getting final rednode config,
- *tree* for getting the tree of rednodes heriarchy.

# GLOBAL OPTIONS

*-v, --verbose*
	Increase the verbosity. Can be repeated. Can optionaly be followed
	by a number terlling the expected verbosity level, 0: EMERGENCY,
	1: ALERT, 2: CRITICAL, 3: ERROR, 4: WARNING, 5: NOTICE, 6: INFO,
	7: DEBUG, 8: TRACE.

*-y, --yaml*
	YAML output where meaningful.

*-l, --log-yaml* _LEVEL_
	Log level of libcyaml when reading and writing YAML data.

*-h, --help*
	Show help and exit

# COMMAND CREATE

The command *create* installs a new rednode. It copies the configuration
files (normal and admin) and set up the rednode status file.

When used with option *--update* the command *create* accepts to update
the configurations of an existing node instead of creating it.

## Synopsys of command create

*redconf* *create* <opts> redpath

## Options of command create

*-a, --alias* _NODE-ALIAS_
	Set the name of the rednode. If not given, the basename of the
	node directory is used.

*-c, --config* _CONFIG-FILE-OR-MODEL_
	Indication for locating the normal configuration file.
	See below *LOCATION OF CONFIG FILES*.

	Default is *leaf*.

*-C, --config-adm* _CONFIG-FILE-OR-MODEL_
	Indication for locating the admin configuration file.
	See below *LOCATION OF CONFIG FILES*.

	Default is *leaf*.

*-t, --templates* _TEMPLATE-DIR_
	Selects an alternate template directory.
	See also environment variable *redpak_TMPL*.

*-u, --update*
	Update an existing rednode.

*-f, --full*
	Shortcut for options *--config full --config-adm full*.

*-l, --leaf*
	Shortcut for options *--config leaf --config-adm leaf*.

*-r, --root*
	Shortcut for options *--config root --config-adm root*.

*-m, --model* _MODEL_
	Shortcut for options *--config MODEL --config-adm MODEL*.

*-n, --node* _REDPATH_
	The path of the node to create. This option is redundant with the
	first positional argument.

*-R, --redroot* _ROOT-PATH_
	The root of rednodes. Default is */*.

*-h, --help*
	Print help on create command usage.


# COMMAND CONFIG

The command *config* outputs the specific configuration of one node.
The specific configuration is the configuration of that node, outside
of any node hierarchy Specific. It means that configuration items coming
from parent nodes hierarchy are not shown by this command (but see
command mergeconfig)

## Synopsys of command config

*redconf* *config* <opts> redpath

## Options of command config

*-r, --redpath* _REDPATH_
	The path of the node whose config is to be displayed.
	This option is redundant with the first positional argument.

*-h, --help*
	Print help on config command usage.


# COMMAND MERGECONFIG

The command *config* outputs the merged configuration of one node.
The merged configuration is the final configuration of that node,
taking into account configurations inherited from parent nodes
hierarchy.

## Synopsys of command mergeconfig

*redconf* *mergeconfig* <opts> redpath

## Options of command mergeconfig

*-e, --expand*
	Expands the expandable variables on output.

*-r, --redpath* _REDPATH_
	The path of the node whose config is to be displayed.
	This option is redundant with the first positional argument.

*-h, --help*
	Print help on mergeconfig command usage.

# COMMAND TREE

The command *tree* displays the tree rednodes starting
from a path in rednode hirarchy.

## Synopsys of command tree

*redconf* *tree* <opts> redpath

## Options of command tree

*-r, --redpath* _REDPATH_
	The path of the node whose config is to be displayed.
	This option is redundant with the first positional argument.

*-h, --help*
	Print help on config command usage.

# LOCATION OF CONFIG FILES

When an argument _ARG_ is given to options *--config* or *--config-adm*,
the below paths are tested in order and the first existing file is used:

. ${PREF}_ARG_
. ${PREF}_ARG_.yaml
. ${PREF}_ARG_-${TYPE}.yaml

With the below values for *${PREF}* and *${TYPE}*:

- If _ARG_ has a slash, it should be a file, so *${PREF}* is the
  empty string. Otherwise, it should be a temùplate indication
  so *${PREF}* is substituted with the path of templates' directory.

- The value of *${TYPE}* is either *normal* or *admin*, depending on
  the option (*--config* or *--config-adm*).

For example, if option given is *--model tiny*, then normal config
is searched in:

. */etc/redpak/templates.d/tiny*
. */etc/redpak/templates.d/tiny.yaml*
. */etc/redpak/templates.d/tiny-admin.yaml*

Adn normal config in:

. */etc/redpak/templates.d/tiny*
. */etc/redpak/templates.d/tiny.yaml*
. */etc/redpak/templates.d/tiny-normal.yaml*


# ENVIRONMENT

The below ernvironment variables can be used.

*REDNODE_TEMPLATE_DIR*
	Location of the template configurations directory.
	Default value: *"/etc/redpak/templates.d"*

# EXAMPLES

This example show the creation 
```
> redconf create -m root /tmp/root -a ROOT
> redconf create /tmp/root/leaf1 -a LEAF-1
> redconf create /tmp/root/leaf1/leaf11 -a LEAF-1-1
> redconf create /tmp/root/leaf1/leaf12 -a LEAF-1-2
> redconf tree /tmp/root
/tmp
  └── root  (ROOT)
      └── leaf1  (LEAF-1)
          ├── leaf11  (LEAF-1-1)
          └── leaf12  (LEAF-1-2)
```

It is possible to get the configuration of one of these node:

```
redconf -y mergeconfig -e /tmp/root/leaf1/leaf12
```

# SEE ALSO

*redwrap*(8), *rednode*(5)

# AUTHOR

IoT.bzh


CHANGELOG
Version 2.3.0 (2024/11/18)

 - fix of cgroup management
 - redconf creates rednodes on demand
 - more tests



red-conf/commands/redconf.c
/*
* Copyright (C) 2022 "IoT.bzh"
*
* Licensed under the Apache License, Version 2.0 (the "License");
* you may not use this file except in compliance with the License.
* You may obtain a copy of the License at
*
*   http://www.apache.org/licenses/LICENSE-2.0
*
* Unless required by applicable law or agreed to in writing, software
* distributed under the License is distributed on an "AS IS" BASIS,
* WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
* See the License for the specific language governing permissions and
* limitations under the License.
*
*/

#include <string.h>
#include <stdio.h>
#include <stdlib.h>

//include subcommands
#include "cmd-config.h"
#include "cmd-mergeconfig.h"
#include "cmd-tree.h"
#include "cmd-create.h"
#include "options.h"

#include "../redconf.h"
#include "../redconf-defaults.h"


typedef struct {
    const char *cmd_name;
    int (*cmd)(const rGlobalConfigT *);
    const char *desc;
} rCommandT;

/***************************************
 **** Global command ****
 **************************************/

//all subcommand of redconf
static const rCommandT commands[] = {
    {"config", config, "get configuration files from nodes"},
    {"mergeconfig", mergeconfig, "get merge configuration files from nodes"},
    {"tree", tree, "get tree nodes"},
    {"create", create, "creates a node"},
    {0, 0, 0}
};

static const rOption globalOptions[] = {
    {{"verbose",    optional_argument,     0, 'v'}, "increase verbosity to info"},
    {{"yaml",       no_argument      ,     0, 'y'}, "yaml output"},
    {{"log-yaml",   required_argument,     0, 'l'}, "yaml parse log level (max=4)"},
    {{"help",       no_argument      ,     0, 'h'}, "print this help"},
    {{0, 0, 0, 0}, 0}
};

/*
* short options with a + at beginning for preserving order
* and stopping at the first argument not being an option
*/
static const char SHORTOPTS[] = "+v::hyl:";

static void globalUsage(const rOption *options, int exitcode) {
    printf("Usage: redconf [OPTION]... COMMAND [OPTION]...\n"
           "redpak configuration commands\n"
           "\n"
           "Commands:\n"
    );
    for (const rCommandT *cmd = commands; cmd->cmd; cmd++)
        printf("\t%-13s\t%s\n", cmd->cmd_name, cmd->desc);

    usageOptions(options);
    exit(exitcode);
}

static int globalParseArgs(int argc, char *argv[], rGlobalConfigT *gConfig) {
    int option, verbosity;
    struct option longOpts [sizeof(struct option) * sizeof(globalOptions) / sizeof(globalOptions[0])];

    setLongOptions(globalOptions, longOpts);
    optind = 0; /* force reset */
    while(1) {
        option = getopt_long(argc, argv, SHORTOPTS, longOpts, NULL);

        // option return short option even when long option is given
        switch (option) {
            case 'v':
                verbosity = GetLogLevel();
                if(!optarg)
                    verbosity++;
                else if (*optarg == 'v') {
                    do { verbosity++; } while(*optarg++ == 'v');
                }
                else
                    verbosity = atoi(optarg);
                SetLogLevel((RedLogLevelE)verbosity);
                break;

            case 'y':
                gConfig->yaml = 1;
                break;

            case 'l':
                setLogYaml(atoi(optarg));
                break;

            case 'h':
                globalUsage(globalOptions, EXIT_SUCCESS);
                break;

            default:
                globalUsage(globalOptions, EXIT_FAILURE);
                break;

            case -1: //no more options
                gConfig->cmd = argv[optind];
                gConfig->sub_argc = argc - optind;
                gConfig->sub_argv = argv + optind;
                optind = 0; /* force reset */
                return 0;
        }
    }
}

int main (int argc, char *argv[]) {
    rGlobalConfigT gConfig = {0};
    if (globalParseArgs(argc, argv, &gConfig) < 0)
        return 1;

    //if no sub command issue
    if(!gConfig.cmd) {
        RedLog(REDLOG_ERROR, "No command options found\n");
    }
    else {
        //call right sub command
        for(const rCommandT *cmd = commands; cmd->cmd_name; cmd++) {
            if (!strcmp(cmd->cmd_name, gConfig.cmd)) {
                return cmd->cmd(&gConfig) < 0 ? EXIT_FAILURE : EXIT_SUCCESS;
            }
        }
        RedLog(REDLOG_ERROR, "Invalid command %s\n", gConfig.cmd);
    }
    globalUsage(globalOptions, EXIT_FAILURE);
    return EXIT_FAILURE;
}